x-pack/plugins/apm/public/context/UrlParamsContext/index.tsx

```diff
@@ -40,58 +35,49 @@ function useUiFilters(urlParams: IUrlParams): UIFilters {
 
 const defaultRefresh = (time: TimeRange) => {};
 
-export function urlParamsReducer(
-  state: IUrlParams = {},
-  action: LocationAction | TimeRangeRefreshAction
-): IUrlParams {
-  switch (action.type) {
-    case LOCATION_UPDATE: {
-      return resolveUrlParams(action.location, state);
-    }
-
-    case TIME_RANGE_REFRESH:
-      return {
-        ...state,
-        start: getParsedDate(action.time.rangeFrom),
-        end: getParsedDate(action.time.rangeTo)
-      };
-
-    default:
-      return state;
-  }
-}
-
 const UrlParamsContext = createContext({
   urlParams: {} as IUrlParams,
   refreshTimeRange: defaultRefresh,
   uiFilters: {} as UIFilters
 });
 
-const UrlParamsProvider: React.FC<{}> = ({ children }) => {
-  const location = useLocation();
-  const [urlParams, dispatch] = useReducer(
-    urlParamsReducer,
-    resolveUrlParams(location, {})
-  );
-  const uiFilters = useUiFilters(urlParams);
+const UrlParamsProvider: React.ComponentClass<{}> = withRouter(
+  ({ location, children }) => {
+    const refUrlParams = useRef(resolveUrlParams(location, {}));
 
-  function refreshTimeRange(time: TimeRange) {
-    dispatch({ type: TIME_RANGE_REFRESH, time });
-  }
+    const [, forceUpdate] = useState('');
 
-  useEffect(
-    () => {
-      dispatch({ type: LOCATION_UPDATE, location });
-    },
-    [location]
-  );
+    const urlParams = resolveUrlParams(location, {
+      start: refUrlParams.current.start,
+      end: refUrlParams.current.end,
+      rangeFrom: refUrlParams.current.rangeFrom,
+      rangeTo: refUrlParams.current.rangeTo
+    });
 
-  return (
-    <UrlParamsContext.Provider
-      children={children}
-      value={{ urlParams, refreshTimeRange, uiFilters }}
-    />
-  );
-};
+    refUrlParams.current = urlParams;
+
+    const refreshTimeRange = useCallback(
+      (timeRange: TimeRange) => {
+        refUrlParams.current = {
+          ...refUrlParams.current,
+          start: getParsedDate(timeRange.rangeFrom),
+          end: getParsedDate(timeRange.rangeTo)
+        };
+
+        forceUpdate(uniqueId());
+      },
+      [forceUpdate]
+    );
+
+    const uiFilters = useUiFilters(urlParams);
+
+    return (
+      <UrlParamsContext.Provider
+        children={children}
+        value={{ urlParams, refreshTimeRange, uiFilters }}
```

Comment:
What do you think about wrapping this in `useMemo` so it doesn't compute a new object on every render?

Reply:
Aye, will fix.